splito/lohi/_lo.py

```diff
@@ -0,0 +1,157 @@
+        """
+        if verbose:
+            progress_bar = tqdm(total=self.max_clusters, desc="Collecting clusters")
+
+        # At first, all the nodes are in the train set. Some will be moved to the list of clusters.
+        train_nodes = np.array(range(len(smiles)))
+
+        train_fps = dm.parallelized(
+            functools.partial(dm.to_fp, as_array=False, radius=2, nBits=1024),
+            smiles,
+            n_jobs=n_jobs,
+        )
+        all_clusters_nodes = []  # the test clusters of nodes
+
+        while len(all_clusters_nodes) < self.max_clusters:
+            total_neighbours, stds = self._get_neighborhood(train_fps, values)
+            central_idx = self._get_central_idx(total_neighbours, stds)
+            if central_idx is None:
+                break  # there are no more clusters
+
+            cluster_idx = self._collect_cluster(central_idx, train_fps)
+
+            # Save the cluster nodes
+            all_clusters_nodes.append(train_nodes[cluster_idx])
+
+            # Remove neighbours of the cluster from the rest of nodes
+            nearest_sim = self._get_nearest_sim(train_fps, cluster_idx)
+            rest_idx = []
+            for idx, sim in enumerate(nearest_sim):
+                if sim < self.threshold:
+                    rest_idx.append(idx)
+            train_nodes = train_nodes[rest_idx]
+            values = values[rest_idx]
+            train_fps = [train_fps[idx] for idx in rest_idx]
+
+            if verbose:
+                progress_bar.update(1)
+        if verbose:
+            progress_bar.close()
+            logger.info(f"Found {len(all_clusters_nodes)} clusters.")
+        return train_nodes, all_clusters_nodes
+
+    def _get_neighborhood(self, train_fps, values):
+        """
+        For each node find number of neighbours and std of their values.
+        """
+        total_neighbours = []
+        stds = []
+        for fps in train_fps:
+            sims = DataStructs.BulkTanimotoSimilarity(fps, train_fps)
+            is_neighbor = np.array(sims) > self.threshold
+            total_neighbours.append(is_neighbor.sum())
+            stds.append(values[is_neighbor].std())
+
+        total_neighbours = np.array(total_neighbours)
+        stds = np.array(stds)
+        return total_neighbours, stds
+
+    def _get_central_idx(self, total_neighbours, stds):
+        """
+        Find the most distant cluster and return the index of its centroid.
+        """
+        central_idx = None
+        least_neighbours = max(total_neighbours)
+        for idx, n_neighbours in enumerate(total_neighbours):
+            if n_neighbours > self.min_cluster_size:
+                if n_neighbours < least_neighbours:
+                    if stds[idx] >= self.std_threshold:
+                        least_neighbours = n_neighbours
+                        central_idx = idx
+        return central_idx
+
+    def _collect_cluster(self, central_idx, train_fps):
```

Comment:
FYI, `datamol` offers clustering function `datamo.cluster_mols`.

Reply:
Yes, Cas suggested it as well. I checked it, but it differs in many ways, so I can't replace my `_collect_cluster`.